### PR TITLE
Consistently use Template#key when evaluating templates in backends

### PR DESCRIPTION
**Pull request by robhanlon22**
When I reintroduced the Template class, I added both the key and text of the message that was looked up at Template instantiation as attributes of the Template class. This led to an inconsistency in how message parameter interpolation is implemented—while the I18n backend just uses the key and looks up the message again, the YAML backend would use the text. This PR makes the approach consistent, and has both message backends look up the message again at the time that interpolation occurs.

### CODE REVIEW

File: lib/dry/schema/messages/yaml.rb [hunk 1]
```diff
@@ -146,31 +141,49 @@ def prepare
...
       private
 
-      def evaluation_context(template)
-        cache.fetch_or_store(template.text) do
-          tokens = template.text.scan(TOKEN_REGEXP).flatten(1).map(&:to_sym).to_set
-          text = template.text.gsub('%', '#')
+      # @api private
+      def evaluate_template(template)
+        result = get(template.key, template.options)
+        return unless result
```
Comment (robhanlon22) [inline lib/dry/schema/messages/yaml.rb]: @solnic do you think this should `raise`?

File: lib/dry/schema/messages/template.rb [hunk 2]
```diff
@@ -10,15 +11,15 @@ module Messages
...
 
         option :messages
         option :key
-        option :text
         option :options
 
         # @api private
         def data(input = EMPTY_HASH)
-          messages.pruned_data(self, **options, **input)
+          messages.interpolatable_data(self, **options, **input)
```
Comment (robhanlon22) [inline lib/dry/schema/messages/template.rb]: I thought that this name (while long) made things more self-documentaing

File: lib/dry/schema/messages/yaml.rb (hunk 1 above)
Comment (robhanlon22) [reply lib/dry/schema/messages/yaml.rb]: Decided to raise a `KeyError` instead of returning `nil`, because this should only happen in exceptional cases

**Comment by solnic** [comment]
@robhanlon22 it's not immediately obvious to me what this PR fixes/improves - could you summarize this somehow please?r

**Comment by robhanlon22** [comment]
@solnic clarified
